subsys/shell/shell_uart.c

```diff
@@ -27,62 +27,68 @@ SHELL_DEFINE(shell_uart, CONFIG_SHELL_PROMPT_UART, &shell_transport_uart,
 	     SHELL_FLAG_OLF_CRLF);
 
 #ifdef CONFIG_SHELL_BACKEND_SERIAL_INTERRUPT_DRIVEN
+static inline bool smp_rx_byte(const struct shell_uart *sh_uart, uint8_t byte)
+{
+#ifdef CONFIG_MCUMGR_SMP_SHELL
+	return smp_shell_rx_byte(&sh_uart->ctrl_blk->smp, byte);
+#else
+	return false;
+#endif
+}
+
 static void uart_rx_handle(struct device *dev,
 			   const struct shell_uart *sh_uart)
 {
 	uint8_t *data;
+	uint8_t dummy;
 	uint32_t len;
+	uint32_t to_read;
 	uint32_t rd_len;
+	uint32_t rd_idx;
+	uint32_t uart_idx;
 	bool new_data = false;
 
 	do {
 		len = ring_buf_put_claim(sh_uart->rx_ringbuf, &data,
 					 sh_uart->rx_ringbuf->size);
-
 		if (len > 0) {
-			rd_len = uart_fifo_read(dev, data, len);
-#ifdef CONFIG_MCUMGR_SMP_SHELL
-			/* Divert bytes from shell handling if it is
-			 * part of an mcumgr frame.
-			 */
-			size_t i;
-
-			for (i = 0; i < rd_len; i++) {
-				if (!smp_shell_rx_byte(&sh_uart->ctrl_blk->smp,
-						       data[i])) {
-					break;
-				}
-			}
+			to_read = len;
```

Comment:
You do not need `to_read`, you may reuse `len` as max length of data to read.
You only use len to figure out whether ring buffer has been successfully obtained and requires 'finishing' (line 77) which you may check by simply comparing `data != &dummy`.